README.md
# helm-migemo.el [![melpa badge][melpa-badge]][melpa-link] [![melpa stable badge][melpa-stable-badge]][melpa-stable-link]

[Migemo](https://github.com/emacs-jp/migemo) plug-in for helm


## Screenshot

![helm-migemo](image/helm-migemo.gif)


## Installation

`helm-migemo` is available on [MELPA][melpa-link] and [MELPA-STABLE][melpa-stable-link].

You can install `helm-migemo` with the following command.

<kbd>M-x package-install [RET] helm-migemo [RET]</kbd>

[melpa-link]: https://melpa.org/#/helm-migemo
[melpa-stable-link]: https://stable.melpa.org/#/helm-migemo
[melpa-badge]: https://melpa.org/packages/helm-migemo-badge.svg
[melpa-stable-badge]: https://stable.melpa.org/packages/helm-migemo-badge.svg


Changes
History: helm-migemo.el

Revision 1.22  2015-10-02 15:20:00  syohex
 - Fix issue for candidates-in-buffer(#6, #7 Reported by tam17aki)

Revision 1.21  2015-10-02 00:14:00  syohex
 - Move repository
 - Clean up code
 - Update dependencies

Revision 1.20  2013-08-20 20:01:12  myuhe
 - Shutup byte-compiler

Revision 1.19  2012-08-18 02:16:22  myuhe
 - Port to helm.

Revision 1.18  2009-06-07 17:52:22  rubikitch
 - New macro `helm-migemize-command'.

Revision 1.17  2009/06/04 20:32:00  rubikitch
 - migemo is soft-required now; this file has no effect unless migemo is installed.

Revision 1.16  2008/10/03 20:43:18  rubikitch
 - Use with helm-match-plugin.el

Revision 1.15  2008/10/03 20:01:46  rubikitch
 - refactoring

Revision 1.14  2008/08/25 08:29:02  rubikitch
 - `helm-migemo': helm-args

Revision 1.13  2008/08/24 20:39:53  rubikitch
 - prevent the unit test from being byte-compiled.

Revision 1.12  2008/08/24 18:01:25  rubikitch

Revision 1.11  2008/08/24 08:23:30  rubikitch
 - Rename `helm-candidates-buffer' -> `helm-candidate-buffer'

Revision 1.10  2008/08/24 01:54:21  rubikitch
 - migemo attribute

Revision 1.9  2008/08/19 21:38:09  rubikitch
 - match attribute bug fix

Revision 1.8  2008/08/19 21:30:29  rubikitch
 - plug-in

Revision 1.7  2008/08/10 22:45:02  rubikitch
 - Bug info

Revision 1.6  2008/08/08 03:40:51  rubikitch
 - require migemo

Revision 1.5  2008/08/08 03:38:34  rubikitch
 - add search attribute
 - unit tests

Revision 1.4  2007/12/26 08:36:01  rubikitch
 - changed match priority

Revision 1.3  2007/12/25 19:55:59  rubikitch
 - patch is not needed anymore.

Revision 1.2  2007/12/25 13:05:46  rubikitch
 - speed up by memoization

Revision 1.1  2007/12/25 12:03:25  rubikitch
 - Initial revision


helm-migemo-test.el
;;;; unit test
;; (install-elisp "http://www.emacswiki.org/cgi-bin/wiki/download/el-expectations.el")
;; (install-elisp "http://www.emacswiki.org/cgi-bin/wiki/download/el-mock.el")
(dont-compile
  (when (fboundp 'expectations)
    (expectations
      (desc "match")
      (expect '(("TEST" ("日本語")))
        (let ((helm-use-migemo t))
          (helm-test-candidates
           '(((name . "TEST")
              (candidates "日本語")))
           "nihongo"
           '(helm-compile-source--migemo))))
      (desc "candidates buffer")
      (expect '(("TEST" ("日本語")))
        (let ((helm-use-migemo t))
          (helm-test-candidates
           '(((name . "TEST")
              (init
               . (lambda () (with-current-buffer (helm-candidate-buffer 'global)
                              (insert "日本語\n"))))
              (candidates-in-buffer)))
           "nihongo"
           '(helm-compile-source--candidates-in-buffer
             helm-compile-source--migemo))))
      (desc "migemo attribute")
      (expect '(("TEST" ("日本語")))
        (let ((helm-use-migemo nil))
          (helm-test-candidates
           '(((name . "TEST")
              (candidates "日本語")
              (migemo)))
           "nihongo"
           '(helm-compile-source--migemo))))
      (expect '(("TEST" ("日本語")))
        (let ((helm-use-migemo nil))
          (helm-test-candidates
           '(((name . "TEST")
              (init
               . (lambda () (with-current-buffer (helm-candidate-buffer 'global)
                              (insert "日本語\n"))))
              (candidates-in-buffer)
              (migemo)))
           "nihongo"
           '(helm-compile-source--candidates-in-buffer
             helm-compile-source--migemo))))
      (desc "search-from-end attribute")
      
      (expect '(("FOO" ("日本語入力" "日本語会話")))
        (let ((helm-use-migemo nil))
          (helm-test-candidates '(((name . "FOO")
                                       (init
                                        . (lambda ()
                                            (with-current-buffer (helm-candidate-buffer 'global)
                                              (insert "日本語会話\n日本語入力\n"))))
                                       (candidates-in-buffer)
                                       (migemo)
                                       (search-from-end)))
                                    "nihongo"
                                    '(helm-compile-source--candidates-in-buffer
                                      helm-compile-source--migemo))))
      (expect '(("FOO" ("日本語入力" "日本語会話")))
        (let ((helm-use-migemo t))
          (helm-test-candidates '(((name . "FOO")
                                       (init
                                        . (lambda ()
                                            (with-current-buffer (helm-candidate-buffer 'global)
                                              (insert "日本語会話\n日本語入力\n"))))
                                       (candidates-in-buffer)
                                       (search-from-end)))
                                    "nihongo"
                                    '(helm-compile-source--candidates-in-buffer
                                      helm-compile-source--migemo))))
      (desc "with helm-match-plugin")
      
      (expect '(("FOO" ("日本語入力")))
        (let ((helm-use-migemo nil))
          (helm-test-candidates '(((name . "FOO")
                                       (init
                                        . (lambda ()
                                            (with-current-buffer (helm-candidate-buffer 'global)
                                              (insert "日本語会話\n日本語入力\n"))))
                                       (candidates-in-buffer)
                                       (migemo)))
                                    "nihongo nyuuryoku"
                                    '(helm-compile-source--candidates-in-buffer
                                      helm-compile-source--match-plugin
                                      helm-compile-source--migemo))))
      (expect '(("FOO" ("日本語入力")))
        (let ((helm-use-migemo t))
          (helm-test-candidates '(((name . "FOO")
                                       (init
                                        . (lambda ()
                                            (with-current-buffer (helm-candidate-buffer 'global)
                                              (insert "日本語会話\n日本語入力\n"))))
                                       (candidates-in-buffer)))
                                    "nihongo nyuuryoku"
                                    '(helm-compile-source--candidates-in-buffer
                                      helm-compile-source--match-plugin
                                      helm-compile-source--migemo))))
      (expect '(("FOO" ("日本語入力")))
        (let ((helm-use-migemo nil))
          (helm-test-candidates '(((name . "FOO")
                                       (init
                                        . (lambda ()
                                            (with-current-buffer (helm-candidate-buffer 'global)
                                              (insert "日本語会話\n日本語入力\n"))))
                                       (candidates-in-buffer)
                                       (search-from-end)
                                       (migemo)))
                                    "nihongo nyuuryoku"
                                    '(helm-compile-source--candidates-in-buffer
                                      helm-compile-source--match-plugin
                                      helm-compile-source--migemo))))
      (expect '(("FOO" ("日本語入力")))
        (let ((helm-use-migemo t))
          (helm-test-candidates '(((name . "FOO")
                                       (init
                                        . (lambda ()
                                            (with-current-buffer (helm-candidate-buffer 'global)
                                              (insert "日本語会話\n日本語入力\n"))))
                                       (candidates-in-buffer)
                                       (search-from-end)))
                                    "nihongo nyuuryoku"
                                    '(helm-compile-source--candidates-in-buffer
                                      helm-compile-source--match-plugin
                                      helm-compile-source--migemo))))
      (expect '(("TEST" ("日本語入力")))
        (let ((helm-use-migemo nil))
          (helm-test-candidates
           '(((name . "TEST")
              (migemo)
              (candidates "日本語入力")))
           "nihongo nyuuryoku"
           '(helm-compile-source--match-plugin helm-compile-source--migemo))))
      (expect '(("TEST" ("日本語入力")))
        (let ((helm-use-migemo t))
          (helm-test-candidates
           '(((name . "TEST")
              (candidates "日本語入力")))
           "nihongo nyuuryoku"
           '(helm-compile-source--match-plugin helm-compile-source--migemo))))
      )))


helm-migemo.el
;;; helm-migemo.el --- Migemo plug-in for helm -*- lexical-binding: t -*-

;; Copyright (C) 2007-2012 rubikitch
;; Copyright (C) 2015 Yuhei Maeda <yuhei.maeda_at_gmail.com>
;; Author: rubikitch <rubikitch@ruby-lang.org>
;; Maintainer: Yuhei Maeda <yuhei.maeda_at_gmail.com>
;; Version: 1.22
;; Package-Requires: ((helm-core "1.7.8") (migemo "1.9") (cl-lib "0.5"))
;; Created: 2009-04-13
;; Keywords: matching, convenience, tools, i18n
;; URL: https://github.com/emacs-jp/helm-migemo

;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:
;;
;; Migemo extension of `helm'. Use `helm-migemo' instead of
;; `helm'. If `helm-migemo' is invoked with prefix argument,
;; `helm' is migemo-ized. This means that pattern matching of
;; `helm' candidates is done by migemo-expanded `helm-pattern'.

;;; Commands:
;;
;; Below are complete command list:
;;
;;  `helm-migemo'
;;    `helm' with migemo extension.
;;
;;; Customizable Options:
;;
;; Below are customizable option list:
;;

;; If you want to use migemo search source-locally, add (migemo) to
;; the source. It sets match and search attribute appropriately for
;; migemo.

;;; Setting:

;; (require 'helm-migemo)
;; (define-key global-map [(control ?:)] 'helm-migemo)

;;; Bug:

;; Simultaneous use of (candidates-in-buffer), (search
;; . migemo-forward) and (delayed) scrambles *helm* buffer. Maybe
;; because of collision of `migemo-process' and `run-with-idle-timer'

;;; Code:

(require 'cl-lib)
(require 'helm)
(require 'migemo nil t)

(defvar helm-use-migemo nil
  "[Internal] If non-nil, `helm' is migemo-ized.")

;;;###autoload
(defun helm-migemo (with-migemo &rest helm-args)
  "`helm' with migemo extension.
With prefix arugument, `helm-pattern' is migemo-ized, otherwise normal `helm'."
  (interactive "P")
  (let ((helm-use-migemo with-migemo))
    (apply 'helm helm-args)))

(defvar helm-previous-migemo-info '("" . "")
  "[Internal] Previous migemo query for helm-migemo.")
(cl-defun helm-string-match-with-migemo (str &optional (pattern helm-pattern))
  "Migemo version of `string-match'."
  (unless (string= pattern (car helm-previous-migemo-info))
    (setq helm-previous-migemo-info (cons pattern (migemo-get-pattern pattern))))
  (string-match (cdr helm-previous-migemo-info) str))

  (cl-defun helm-mp-3migemo-match (str &optional (pattern helm-pattern))
    (cl-loop for (pred . re) in (helm-mm-3-get-patterns pattern)
             always (funcall pred (helm-string-match-with-migemo str re))))
  (defun helm-mp-3migemo-search (pattern &rest _ignore)
    (helm-mm-3-search-base pattern 'migemo-forward 'migemo-forward))
  (defun helm-mp-3migemo-search-backward (pattern &rest _ignore)
    (helm-mm-3-search-base pattern 'migemo-backward 'migemo-backward))
;; (helm-string-match-with-migemo "日本語入力" "nihongo")
;; (helm-string-match-with-migemo "日本語入力" "nyuuryoku")
;; (helm-mp-3migemo-match "日本語入力" "nihongo nyuuryoku")
(defun helm-compile-source--migemo (source)
  (if (not (featurep 'migemo))
      source
    (let* ((match-identity-p
            (or (assoc 'candidates-in-buffer source)
                (equal '(identity) (assoc-default 'match source))))

           (matcher 'helm-mp-3migemo-match)
           (searcher (if (assoc 'search-from-end source)

                           'helm-mp-3migemo-search-backward
                         'helm-mp-3migemo-search)))
      (cond (helm-use-migemo
             `((delayed)
               (search ,@(assoc-default 'search source) ,searcher)
               ,(if match-identity-p
                    '(match identity)
                  `(match ,matcher
                          ,@(assoc-default 'match source)))
               ,@source))
            ((assoc 'migemo source)
             `((search ,searcher)
               ,(if match-identity-p
                    '(match identity)
                  `(match ,matcher))
               ,@source))
            (t source)))))
(add-to-list 'helm-compile-source-functions 'helm-compile-source--migemo t)

(defvar helm-migemize-command-idle-delay 0.1
  "`helm-idle-delay' for migemized command.")
(defmacro helm-migemize-command (command)
  "Use migemo in COMMAND when selectiong candidate by `helm'.
Bind `helm-use-migemo' = t in COMMAND.
`helm-migemize-command-idle-delay' is used instead of  `helm-idle-delay'."
  `(defadvice ,command (around helm-use-migemo activate)
     (let ((helm-use-migemo t)
           (helm-idle-delay helm-migemize-command-idle-delay))
       ad-do-it)))

(with-eval-after-load "helm-migemo"
  ;; redefinition
  (defun helm-compile-source--candidates-in-buffer (source)
    (helm-aif (assoc 'candidates-in-buffer source)
        (append source
                `((candidates
                   . ,(or (cdr it)
                          (lambda ()
                            ;; Do not use `source' because other plugins
                            ;; (such as helm-migemo) may change it
                            (helm-candidates-in-buffer (helm-get-current-source)))))
                  (volatile) (match identity)))
      source)))

(provide 'helm-migemo)

;; How to save (DO NOT REMOVE!!)
;; (progn (magit-push) (emacswiki-post "helm-migemo.el"))
;;; helm-migemo.el ends here
